<a href="https://colab.research.google.com/github/barkle2/Programming/blob/master/%5BPython%5D%20Tutorial/kosis_API.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 통계청 OPEN API 활용 예제

### 1. 인증키 받기
- 가장 먼저 해야하는 일은 KOSIS 공유서비스(https://kosis.kr/openapi/index/index.jsp) 사이트에 가입하고, 인증키를 받는 것입니다.
- 인증키 유효기간은 2년입니다.

In [1]:
key = "MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ="

### 2. 통계목록 확인
- 요청 URL: http://kosis.kr/openapi/statisticsList.do?method=getList
- 요청 변수(String)
  - apiKey : 인증키
  - vwCd
    - MT_ZTITLE : 국내통계 주제별
    - MT_OTITLE : 국내통계 기관별
  - parentListId
    - MT_ZTITLE : 노동(D)
    - MT_OTITLE : 통계청(101), 고용노동부(118)
  - format: json

- API 호출을 위한 Query 문 생성

In [20]:
query = "https://kosis.kr/openapi/statisticsList.do?method=getList&apiKey={}&vwCd={}&parentListId={}&format=json&jsonVD=Y"
vwCd = "MT_ZTITLE"
parentListId = "D"

query = query.format(key, vwCd, parentListId)

- requests library를 활용하여 API 호출

In [3]:
import requests
import pandas as pd
from google.colab import files
import time, random

- '국내통계 주제별' 중 '노동'에 해당하는 통계목록은 다음과 같다.

In [29]:
r = requests.get(query)
r.content
data = r.json()
df = pd.DataFrame(data)
df

LIST_NM      VW_CD     VW_NM LIST_ID
0   경력단절여성등의경제활동실태조사  MT_ZTITLE  국내통계 주제별     D_1
1           경제활동인구조사  MT_ZTITLE  국내통계 주제별     D_2
2            고령자고용현황  MT_ZTITLE  국내통계 주제별     D_3
3             고용보험통계  MT_ZTITLE  국내통계 주제별     D_4
4          고용허가제고용동향  MT_ZTITLE  국내통계 주제별     D_5
5        고용형태별근로실태조사  MT_ZTITLE  국내통계 주제별     D_6
6           구직급여신청동향  MT_ZTITLE  국내통계 주제별     D_7
7             근로환경조사  MT_ZTITLE  국내통계 주제별     D_8
8         기업직업훈련실태조사  MT_ZTITLE  국내통계 주제별     D_9
9       기업체장애인고용실태조사  MT_ZTITLE  국내통계 주제별    D_10
10           노동생산성지수  MT_ZTITLE  국내통계 주제별    D_11
11            노사분규통계  MT_ZTITLE  국내통계 주제별    D_12
12       대졸자직업이동경로조사  MT_ZTITLE  국내통계 주제별    D_13
13          박사인력활동조사  MT_ZTITLE  국내통계 주제별    D_14
14     사업체기간제근로자현황조사  MT_ZTITLE  국내통계 주제별    D_15
15          사업체노동력조사  MT_ZTITLE  국내통계 주제별    D_16
16         사업체노동실태현황  MT_ZTITLE  국내통계 주제별    D_17
17      산업기술인력수급실태조사  MT_ZTITLE  국내통계 주제별    D_18
18            산재보험통계  MT_ZTITLE  국내통계 주제별    D_19
19          산재보험패널조사  MT_ZTITLE  국내통계 주제별    D_20
20         여성관리자패널조사  MT_ZTITLE  국내통계 주제별    D_21
21      워크넷구인구직및취업동향  MT_ZTITLE  국내통계 주제별    D_22
22      이민자체류실태및고용조사  MT_ZTITLE  국내통계 주제별    D_23
23           일자리이동통계  MT_ZTITLE  국내통계 주제별   D_001
24           일자리행정통계  MT_ZTITLE  국내통계 주제별    D_24
25     임금근로일자리동향행정통계  MT_ZTITLE  국내통계 주제별    D_25
26            육아휴직통계  MT_ZTITLE  국내통계 주제별   B_001
27          작업환경실태조사  MT_ZTITLE  국내통계 주제별    D_26
28       장애인경제활동실태조사  MT_ZTITLE  국내통계 주제별    D_27
29         장애인고용패널조사  MT_ZTITLE  국내통계 주제별    D_28
30      장애인구인구직및취업동향  MT_ZTITLE  국내통계 주제별    D_29
31         장애인의무고용현황  MT_ZTITLE  국내통계 주제별    D_30
32        전국노동조합조직현황  MT_ZTITLE  국내통계 주제별    D_31
33           지역별고용조사  MT_ZTITLE  국내통계 주제별    D_32
34       직종별사업체노동력조사  MT_ZTITLE  국내통계 주제별    D_33
35            청년패널조사  MT_ZTITLE  국내통계 주제별    D_34
36          한국노동패널조사  MT_ZTITLE  국내통계 주제별    D_35

- 엑셀 파일로 저장한 후 다운로드

In [31]:
df.to_excel('노동 통계목록.xlsx')
files.download('노동 통계목록.xlsx')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

- 경제활동인구조사 통계 내에도 여러 개의 세부 통계목록이 존재한다.

In [32]:
query = "https://kosis.kr/openapi/statisticsList.do?method=getList&apiKey={}&vwCd={}&parentListId={}&format=json&jsonVD=Y"
vwCd = "MT_ZTITLE"
parentListId = "D_2"

query = query.format(key, vwCd, parentListId)

In [33]:
r = requests.get(query)
r.content
data = r.json()
df = pd.DataFrame(data)
df

LIST_NM  ...  LIST_ID
0             경제활동인구총괄(공식 실업률, 구직기간 4주기준, 1999.6~현재)  ...      B11
1                             계절조정 경제활동인구(1999.6~현재)  ...      B12
2                                       취업자(1963~현재)  ...      B17
3    실업자(구직기간 4주 기준, 1999.5월 이전자료는 경제활동인구총괄 1주기준 참조)  ...      B16
4   실업률(구직기간 4주 기준, 1999. 5월 이전자료는 경제활동인구총괄 1주기준 참조)  ...      B15
5           비경제활동인구(1999. 5월 이전자료는 경제활동인구총괄 1주기준 참조)  ...      B14
6                  경제활동인구총괄(구직기간 1주기준, 1963~2014.12)  ...      B13
7                                         근로형태별 부가조사  ...  101_B1A
8                                           청년층 부가조사  ...   B1A_18
9                                           고령층 부가조사  ...   B1A_20
10                                        비임금근로 부가조사  ...   B1A_22
11                                비임금근로 부가조사(농림어업제외)  ...   B1A_24
12                                      비경제활동인구 부가조사  ...   B1A_26
13                         시계열 보정(2018.2.14 공표) 前 자료  ...  B1A_101

[14 rows x 4 columns]

In [ ]:
df.to_excel('경제활동인구조사 통계목록.xlsx')
files.download('경제활동인구조사 통계목록.xlsx')

- 가장 많이 사용되는 '경제활동인구총괄(공식 실업률, 구직기간 4주기준, 1999.6~현재)' 표의 세부 목록을 확인한다.

In [34]:
query = "https://kosis.kr/openapi/statisticsList.do?method=getList&apiKey={}&vwCd={}&parentListId={}&format=json&jsonVD=Y"
vwCd = "MT_ZTITLE"
parentListId = "B11"

query = query.format(key, vwCd, parentListId)

In [35]:
r = requests.get(query)
r.content
data = r.json()
df = pd.DataFrame(data)
df

TBL_NM        TBL_ID      VW_CD     VW_NM ORG_ID
0            성별 경제활동인구 총괄   DT_1DA7001S  MT_ZTITLE  국내통계 주제별    101
1           연령별 경제활동인구 총괄   DT_1DA7002S  MT_ZTITLE  국내통계 주제별    101
2            교육정도별 경제활동인구   DT_1DA7003S  MT_ZTITLE  국내통계 주제별    101
3        행정구역(시도)별 경제활동인구   DT_1DA7004S  MT_ZTITLE  국내통계 주제별    101
4            성/연령별 경제활동인구   DT_1DA7012S  MT_ZTITLE  국내통계 주제별    101
5          성/교육정도별 경제활동인구   DT_1DA7013S  MT_ZTITLE  국내통계 주제별    101
6      행정구역(시도)/성별 경제활동인구   DT_1DA7014S  MT_ZTITLE  국내통계 주제별    101
7     행정구역(시도)/연령별 경제활동인구   DT_1DA7015S  MT_ZTITLE  국내통계 주제별    101
8   행정구역(시도)/교육정도별 경제활동인구   DT_1DA7016S  MT_ZTITLE  국내통계 주제별    101
9              고용보조지표(전체)   DT_1DA7300S  MT_ZTITLE  국내통계 주제별    101
10    고용보조지표(청년층, 15~29세)  DT_1DA7300AS  MT_ZTITLE  국내통계 주제별    101

In [ ]:
df.to_excel('경제활동인구조사-경제활동인구총괄 통계목록.xlsx')
files.download('경제활동인구조사-경제활동인구총괄 통계목록.xlsx')

- '연령별 경제활동인구 총괄' 표의 통계자료를 조회하기 위해 필요한 테이블 ID(TBL_ID:DT_1DA7002S)와 기관 ID(ORG_ID:101)를 얻었다.

In [40]:
tbl_id = 'DT_1DA7002S'
org_id = '101'

## 3. 통계자료 조회

### 3-1. 최신 시점 찾기
- 현재 기준으로 가장 최신 통계가 언제의 통계인지 확인한다.
- '횡단면 조회' 자료를 받아 최신 통계 시점을 확인한다.
- 통계청 OPEN API 사이트에서 횡단면 조회 query 예제 복사

In [45]:
# 0:apiKey, 1:itmId, 2:objL1, 3:prdSe, 4:loadGubun, 5:ordId, 6:tblId
query = "https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey={0}&itmId={1}+&objL1={2}+&objL2=&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe={3}&newEstPrdCnt=1&loadGubun={4}&orgId={5}&tblId={6}"

- 통계청 OPEN API 사이트에서 항목, 분류 value 찾기

In [ ]:
# 항목
itm_id = {
    "인구":"T10",
    "경제활동인구":"T20",
    "취업자":"T30",
    "실업자":"T40",
    "비경제활동인구":"T50",
    "경제활동참가율":"T60",
    "실업률":"T80",
    "고용률":"T90"
}

# 1단계 분류
obj_lev1 = {
     "15세 이상 전체": "00",
     "15 - 19세": "10",
     "20 - 29세": "20",
     "30 - 39세": "30",
     "40 - 49세": "40",
     "50 - 59세": "50",
     "60세 이상": "60",
     "15 - 64세": "63",
     "15 - 24세": "70",
     "15 - 29세": "75"
}

# 2단계 분류
obj_lev2 = {
     "15세 이상 전체": "00",
     "15 - 19세": "10",
     "20 - 29세": "20",
     "20 - 24세": "201",
     "25 - 29세": "202",
     "30 - 39세": "30",
     "30 - 34세": "301",
     "35 - 39세": "302",
     "40 - 49세": "40",
     "40 - 44세": "401",
     "45 - 49세": "402",
     "50 - 59세": "50",
     "50 - 54세": "501",
     "55 - 59세": "502",
     "60세 이상": "60",
     "60 - 64세": "601",
     "65세 이상": "602",
     "15 - 64세": "63",
     "15 - 24세": "70",
     "15 - 29세": "75"
}

- 그 밖에 필요한 변수는 직접 입력

In [46]:
prd_se = 'M'
load_gubun = '2'

query = query.format(key, itm_id['인구'], obj_lev1['15세 이상 전체'], prd_se, load_gubun, org_id, tbl_id)
query

'https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&itmId=T10+&objL1=00+&objL2=&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe=M&newEstPrdCnt=1&loadGubun=2&orgId=101&tblId=DT_1DA7002S'

- '횡단면 조회'의 PRD_DE 값이 가장 최근 시점(end_prd_de)이 된다.

In [47]:
r = requests.get(query)
r.content
data = r.json()
df = pd.DataFrame(data)
df

TBL_NM  PRD_DE       TBL_ID ITM_NM  ... PRD_SE  C1      C1_NM C1_NM_ENG
0  연령별 경제활동인구 총괄  202011  DT_1DA7002S     인구  ...      M  00  15세 이상 전체     Total

[1 rows x 16 columns]

In [64]:
end_prd_de = df['PRD_DE'][0]
end_prd_de

'202011'

- 가장 최근 시점을 구했으니 시계열 조회를 한다.
- 통계청 OPEN API 사이트에서 시계열 조회 query 예제 복사

In [65]:
# 0:apiKey, 1:itmId, 2:objL1, 3:objL2, 4:prdSe, 5:startPrdDe, 6:endPrdDe, 7:loadGubun, 8:orgId, 9:tblId
query = "https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey={0}&itmId={1}+&objL1={2}+&objL2={3}&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe={4}&startPrdDe={5}&endPrdDe={6}&loadGubun={7}&orgId={8}&tblId={9}"

start_prd_de = '199906'
load_gubun = '1'
org_id
tbl_id

query.format(key, itm_id['인구'], obj_lev1['15세 이상 전체'], '', prd_se, start_prd_de, end_prd_de, load_gubun, org_id, tbl_id )

'https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&itmId=T10+&objL1=00+&objL2=&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe=M&startPrdDe=199906&endPrdDe=202011&loadGubun=1&orgId=101&tblId=DT_1DA7002S'

In [9]:
for item in itmId:
  url2 = url2.format(key, item)
  r = requests.get(url2)
  data = r.json()
  df = pd.DataFrame(data)
  filename = item+".csv"
  df.to_csv(filename)
  files.download(filename)
  time.sleep(random.uniform(2,10))


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
r = requests.get(url1)
data = r.json()
r.elapsed.total_seconds()

1.293361

In [ ]:
r = requests.get(url2)
data = r.json()
r.elapsed.total_seconds()

1.187988

In [ ]:
df = pd.DataFrame(data)
df[['PRD_DE', 'DT']]


# 노동 - 통계목록
https://kosis.kr/openapi/statisticsList.do?method=getList&apiKey={}&vwCd=MT_ZTITLE&parentListId=D&format=json&jsonVD=Y

# 경제활동인구조사 - 연령별 경제활동인구 총괄 - 199906~202011 - 15세 이상 - 인구
https://kosis.kr/openapi/statisticsData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&format=json&jsonVD=Y&userStatsId=barkle/101/DT_1DA7002S/2/1/20201228141349_1&prdSe=M&startPrdDe=199906&endPrdDe=202011
# 경제활동인구조사 - 연령별 경제활동인구 총괄 - 199906~202011 - 15세 이상 - 경제활동인구
https://kosis.kr/openapi/statisticsData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&format=json&jsonVD=Y&userStatsId=barkle/101/DT_1DA7002S/2/1/20201228141656_1&prdSe=M&startPrdDe=199906&endPrdDe=202011
# 경제활동인구조사 - 연령별 경제활동인구 총괄 - 199906~202011 - 15세 이상 - 취업자
https://kosis.kr/openapi/statisticsData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&format=json&jsonVD=Y&userStatsId=barkle/101/DT_1DA7002S/2/1/20201228142444_1&prdSe=M&startPrdDe=199906&endPrdDe=202011


url1 = ""
url2 = "https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey={}&itmId={}+&objL1=00+&objL2=&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe=M&startPrdDe=199906&endPrdDe=202011&loadGubun=1&orgId=101&tblId=DT_1DA7002S"
https://kosis.kr/openapi/statisticsData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&format=json&jsonVD=Y&userStatsId=barkle/101/DT_1DA7002S/2/1/20201228115448_1&prdSe=M&startPrdDe=199906&endPrdDe=202011

PRD_DE       DT
0   199906    35737
1   199907    35779
2   199908    35815
3   199909    35849
4   199910    35885
5   199911    35916
6   199912    35953
7   200001  35977.8
8   200002    36011
9   200003  36043.8
10  200004  36081.1
11  200005  36126.4
12  200006  36167.3
13  200007  36215.5
14  200008  36259.2
15  200009  36301.1
16  200010  36338.5
17  200011  36372.9
18  200012    36414
19  200101  36426.7

In [ ]:
df['PRD_DE', 'DT']

KeyError: ignored

In [ ]:
columns = [{'name':i, 'id':i} for i in df.columns]
columns

[{'id': 'LIST_NM', 'name': 'LIST_NM'},
 {'id': 'VW_CD', 'name': 'VW_CD'},
 {'id': 'VW_NM', 'name': 'VW_NM'},
 {'id': 'LIST_ID', 'name': 'LIST_ID'}]

In [ ]:
https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&itmId=T10+&objL1=00+&objL2=&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe=M&startPrdDe=199906&endPrdDe=202011&loadGubun=1&orgId=101&tblId=DT_1DA7002S
https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&itmId=T20+&objL1=00+&objL2=&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe=M&startPrdDe=199906&endPrdDe=202011&loadGubun=1&orgId=101&tblId=DT_1DA7002S
https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&itmId=T30+&objL1=00+&objL2=&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe=M&startPrdDe=199906&endPrdDe=202011&loadGubun=1&orgId=101&tblId=DT_1DA7002S
https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&itmId=T40+&objL1=00+&objL2=&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe=M&startPrdDe=199906&endPrdDe=202011&loadGubun=1&orgId=101&tblId=DT_1DA7002S
https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&itmId=T50+&objL1=00+&objL2=&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe=M&startPrdDe=199906&endPrdDe=202011&loadGubun=1&orgId=101&tblId=DT_1DA7002S
https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&itmId=T60+&objL1=00+&objL2=&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe=M&startPrdDe=199906&endPrdDe=202011&loadGubun=1&orgId=101&tblId=DT_1DA7002S
https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&itmId=T80+&objL1=00+&objL2=&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe=M&startPrdDe=199906&endPrdDe=202011&loadGubun=1&orgId=101&tblId=DT_1DA7002S
https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&itmId=T90+&objL1=00+&objL2=&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe=M&startPrdDe=199906&endPrdDe=202011&loadGubun=1&orgId=101&tblId=DT_1DA7002S

https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&itmId=T10+&objL1=10+&objL2=&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe=M&startPrdDe=199906&endPrdDe=202011&loadGubun=1&orgId=101&tblId=DT_1DA7002S
https://kosis.kr/openapi/statisticsData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&format=json&jsonVD=Y&userStatsId=barkle/101/DT_1DA7002S/2/1/20201228113855_1&prdSe=M&startPrdDe=199906&endPrdDe=202011
